### PR TITLE
Kafka 4.1.0 upgrade

### PR DESCRIPTION
**Pull request by robobario**
### Type of change

- Enhancement / new feature

### Description

Upgrade to kafka 4.1.0

Adopts upstream apache JRE-based Kafka images for 4.1.0+ for Testcontainers deployment mode

A notable difference is that we now use the Admin client to create the SASL SCRAM users for both in-vm and testcontainer mode. This avoids having to use internal code when formatting the storage, using a public API to install the users.

Currently we cannot use the native variant of the apache image as there are some problems that prevent the use of SASL authentication.
https://issues.apache.org/jira/browse/KAFKA-19583
https://issues.apache.org/jira/browse/KAFKA-19584

If your test targets older versions of Kafka <4.1.0 then they will use the established `quay.io/ogunalp/kafka-native`.

The apache images are pulled from `mirror.gcr.io/apache/kafka`, this can be overridden with a new environment variable `APACHE_KAFKA_IMAGE_REPO`. The existing `KAFKA_IMAGE_REPO` environment variable will only be applied to older target kafka version using images from `quay.io/ogunalp/kafka-native`.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

File: impl/src/main/java/io/kroxylicious/testing/kafka/clients/CloseableAdmin.java [hunk 1]
```diff
@@ -14,116 +14,7 @@
...
-import org.apache.kafka.clients.admin.RemoveRaftVoterOptions;
-import org.apache.kafka.clients.admin.RemoveRaftVoterResult;
-import org.apache.kafka.clients.admin.RenewDelegationTokenOptions;
-import org.apache.kafka.clients.admin.RenewDelegationTokenResult;
-import org.apache.kafka.clients.admin.UnregisterBrokerOptions;
-import org.apache.kafka.clients.admin.UnregisterBrokerResult;
-import org.apache.kafka.clients.admin.UpdateFeaturesOptions;
-import org.apache.kafka.clients.admin.UpdateFeaturesResult;
-import org.apache.kafka.clients.admin.UserScramCredentialAlteration;
+import org.apache.kafka.clients.admin.*;
```
Comment (robobario) [inline impl/src/main/java/io/kroxylicious/testing/kafka/clients/CloseableAdmin.java:17]: bad IDEA

File: impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java [hunk 2]
```diff
@@ -250,16 +254,27 @@ private KafkaContainer buildKafkaContainer(KafkaClusterConfig.ConfigHolder holde
...
+                .withCopyToContainer(Transferable.of(propertiesToBytes(properties), 0644), "/etc/kafka/docker/server.properties")
                 .withStartupAttempts(CONTAINER_STARTUP_ATTEMPTS)
                 .withMinimumRunningDuration(MINIMUM_RUNNING_DURATION)
                 .withStartupTimeout(STARTUP_TIMEOUT);
 
         if (clusterConfig.isSaslScram() && !clusterConfig.getUsers().isEmpty()) {
+            if (clusterConfig.isKafkaPostVersion4()) {
+                throw new RuntimeException("scram not implemented for apache kafka");
+            }
+            // todo get scram users into apache native
```
Comment (robobario) [inline impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java:277]: wondering, do we need to support the ozan-native-style configuration where we format the storage with scram users, so that nodes can uses scram auth for intercommunication? Should we instead use the admin client to create the scram creds after the cluster is running?

**Comment by robobario** [comment]
mm tests failing with what might be an upstream issue, native image failing to reflectively call the no-args constructor OAuthBearerUnsecuredLoginCallbackHandler. Wonder if it needs to be added to some graal config.


```
2025-08-06 05:14:09 ERROR testcontainers-lifecycle-0 tc.docker.io/apache/kafka-native:latest:549 - Log output from the failed container:
===> User
uid=1000(appuser) gid=1000(appuser) groups=1000(appuser)
===> Setting default values of environment variables if not already set.
===> Configuring ...
===> Launching ... 
WARNING: THIS IS AN EXPERIMENTAL DOCKER IMAGE RECOMMENDED FOR LOCAL TESTING AND DEVELOPMENT PURPOSES.
[...
[2025-08-06 05:14:08,765] ERROR [BrokerServer id=0] Fatal error during broker startup. Prepare to shutdown (kafka.server.BrokerServer)
org.apache.kafka.common.KafkaException: org.apache.kafka.common.KafkaException: Could not find a public no-argument constructor for org.apache.kafka.common.security.oauthbearer.internals.unsecured.OAuthBearerUnsecuredLoginCallbackHandler
	at org.apache.kafka.common.network.SaslChannelBuilder.configure(SaslChannelBuilder.java:184) ~[?:?]
	at org.apache.kafka.common.network.ChannelBuilders.create(ChannelBuilders.java:188) ~[?:?]
	at org.apache.kafka.common.network.ChannelBuilders.serverChannelBuilder(ChannelBuilders.java:105) ~[?:?]
	at kafka.network.Processor.<init>(SocketServer.scala:883) ~[?:?]
	at kafka.network.Acceptor.newProcessor(SocketServer.scala:791) ~[kafka.Kafka:?]
	at kafka.network.Acceptor.$anonfun$addProcessors$1(SocketServer.scala:757) ~[kafka.Kafka:?]
	at scala.collection.immutable.Range.foreach$mVc$sp(Range.scala:190) ~[kafka.Kafka:?]
	at kafka.network.Acceptor.addProcessors(SocketServer.scala:756) ~[kafka.Kafka:?]
	at kafka.network.DataPlaneAcceptor.configure(SocketServer.scala:472) ~[?:?]
	at kafka.network.SocketServer.createDataPlaneAcceptorAndProcessors(SocketServer.scala:222) ~[?:?]
	at kafka.network.SocketServer.$anonfun$new$16(SocketServer.scala:149) ~[?:?]
	at kafka.network.SocketServer.$anonfun$new$16$adapted(SocketServer.scala:149) ~[?:?]
	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:619) ~[kafka.Kafka:?]
	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:617) ~[kafka.Kafka:?]
	at scala.collection.AbstractIterable.foreach(Iterable.scala:935) ~[kafka.Kafka:?]
	at kafka.network.SocketServer.<init>(SocketServer.scala:149) ~[?:?]
	at kafka.server.BrokerServer.startup(BrokerServer.scala:274) ~[?:?]
	at kafka.server.KafkaRaftServer.$anonfun$startup$2(KafkaRaftServer.scala:96) ~[?:?]
	at kafka.server.KafkaRaftServer.$anonfun$startup$2$adapted(KafkaRaftServer.scala:96) ~[?:?]
	at scala.Option.foreach(Option.scala:437) [kafka.Kafka:?]
	at kafka.server.KafkaRaftServer.startup(KafkaRaftServer.scala:96) [kafka.Kafka:?]
	at kafka.Kafka$.main(Kafka.scala:97) [kafka.Kafka:?]
	at kafka.docker.KafkaDockerWrapper$.main(KafkaDockerWrapper.scala:68) [kafka.Kafka:?]
	at kafka.docker.KafkaDockerWrapper.main(KafkaDockerWrapper.scala) [kafka.Kafka:?]
	at java.base/java.lang.invoke.LambdaForm$DMH/sa346b79c.invokeStaticInit(LambdaForm$DMH) [kafka.Kafka:?]
Caused by: org.apache.kafka.common.KafkaException: Could not find a public no-argument constructor for org.apache.kafka.common.security.oauthbearer.internals.unsecured.OAuthBearerUnsecuredLoginCallbackHandler
	at org.apache.kafka.common.utils.Utils.newInstance(Utils.java:403) ~[?:?]
	at org.apache.kafka.common.security.authenticator.LoginManager.<init>(LoginManager.java:65) ~[?:?]
	at org.apache.kafka.common.security.authenticator.LoginManager.acquireLoginManager(LoginManager.java:116) ~[?:?]
	at org.apache.kafka.common.network.SaslChannelBuilder.configure(SaslChannelBuilder.java:170) ~[?:?]
	... 24 more
Caused by: java.lang.NoSuchMethodException: org.apache.kafka.common.security.oauthbearer.internals.unsecured.OAuthBearerUnsecuredLoginCallbackHandler.<init>()
	at java.base/java.lang.Class.checkMethod(DynamicHub.java:1075) ~[kafka.Kafka:?]
	at java.base/java.lang.Class.getConstructor0(DynamicHub.java:1238) ~[kafka.Kafka:?]
	at java.base/java.lang.Class.getDeclaredConstructor(DynamicHub.java:2930) ~[kafka.Kafka:?]
	at org.apache.kafka.common.utils.Utils.newInstance(Utils.java:401) ~[?:?]
	at org.apache.kafka.common.security.authenticator.LoginManager.<init>(LoginManager.java:65) ~[?:?]
	at org.apache.kafka.common.security.authenticator.LoginManager.acquireLoginManager(LoginManager.java:116) ~[?:?]
	at org.apache.kafka.common.network.SaslChannelBuilder.configure(SaslChannelBuilder.java:170) ~[?:?]
	... 24 more
```

File: impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java (hunk 2 above)
Comment (k-wall) [reply impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java]: > Should we instead use the admin client to create the scram creds after the cluster is running?

I think so, let's use the Kafka public APIs everywhere.

**Comment by k-wall** [comment]
> I tried out share groups and found it didn't work neatly out of the box,
> by default the internal topic tries to be configured with replication factor 3 and fails with our single-node default cluster.

You are talking about just "__share_group_state", right?  I think they'll be edge cases associated with add/remove node.   IIRC we had to tune them for those to suit those use-cases.


> I think we should do the same in in-vm mode to make it convenient to test with Share Groups

Why would we have different behaviour for in-VM vs Container?

General point, I'm not against getting the extension to play nicely with the Share Groups and it would be certainly be nice to ready when the Share Groups feature goes GA in Kafka.  But I think a separate issue is the way to go and I don't think it is the highest priority thing right now.

File: impl/src/main/java/io/kroxylicious/testing/kafka/common/KafkaClusterConfig.java [hunk 3]
```diff
@@ -80,6 +81,7 @@ public class KafkaClusterConfig {
     private static final String SCRAM_256_SASL_MECHANISM_NAME = "SCRAM-SHA-256";
     private static final String SCRAM_512_SASL_MECHANISM_NAME = "SCRAM-SHA-512";
     private static final Pattern PRE_KAFKA_39 = Pattern.compile("^3\\.[0-8]\\..*$");
+    private static final Pattern POST_KAFKA_4_1 = Pattern.compile("^4\\.[1-9][0-9]*\\..*$");
```
Comment (k-wall) [inline impl/src/main/java/io/kroxylicious/testing/kafka/common/KafkaClusterConfig.java:84]: Post 4.1 to means means 4.2 onwards.  Do you mean post 4.0?

Comment (robobario) [reply impl/src/main/java/io/kroxylicious/testing/kafka/common/KafkaClusterConfig.java:84]: Maybe inverting it to be like the existing 'pre-3.9' language would be less ambiguous.  So we would apply the old logic to pre-4.1

File: impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java [hunk 4]
```diff
@@ -407,6 +432,11 @@ public synchronized void start() {
                     clusterConfig.getAnonConnectConfigForCluster(buildBrokerListFor(Listener.ANON)),
                     READY_TIMEOUT_SECONDS, TimeUnit.SECONDS,
                     clusterConfig.getBrokersNum());
+            if (clusterConfig.isSaslScram() && !clusterConfig.getUsers().isEmpty()) {
+                if (clusterConfig.isKafkaPostVersion41()) {
```
Comment (k-wall) [inline impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java:436]: Could we switch to this approach wholesale (i mean <4.0)?   Would it be much work. The motivation to to have fewer conditions.

**Comment by robobario** [comment]
> > I tried out share groups and found it didn't work neatly out of the box,
> > by default the internal topic tries to be configured with replication factor 3 and fails with our single-node default cluster.
> 
> You are talking about just "__share_group_state", right? I think they'll be edge cases associated with add/remove node. IIRC we had to tune them for those to suit those use-cases.
> 
> > I think we should do the same in in-vm mode to make it convenient to test with Share Groups
> 
> Why would we have different behaviour for in-VM vs Container?

My  mistake, we control the full set of properties for both strategies, I thought the built in defaults of the container were involved.

> General point, I'm not against getting the extension to play nicely with the Share Groups and it would be certainly be nice to ready when the Share Groups feature goes GA in Kafka. But I think a separate issue is the way to go and I don't think it is the highest priority thing right now.

Will make an issue, I just saw this issue for adding a closeable share consumer and thought it would be an easy thing to include in the 4.1 branch, but testing it I couldn't get it to consume any messages, took quite a while to spot this topic creation failure as you need logs set to DEBUG

File: impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java [hunk 5]
```diff
@@ -463,78 +456,6 @@ private static void trapKafkaSystemExit() {
         return CloseableAdmin.create(clusterConfig.getAnonConnectConfigForCluster(buildBrokersListFor(Listener.ANON)));
     }
 
-    private List<String> buildScramArguments() {
```
Comment (k-wall) [inline impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java:466]: Glad to see all of this die

File: impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java [hunk 6]
```diff
@@ -109,6 +112,7 @@ public class TestcontainersKafkaCluster implements Startable, KafkaCluster, Kafk
     private static final int ZOOKEEPER_PORT = 2181;
 
     private static final String QUAY_KAFKA_IMAGE_REPO = "quay.io/ogunalp/kafka-native";
+    private static final String APACHE_KAFKA_JAVA_IMAGE_REPO = "docker.io/apache/kafka";
```
[the remark this answers is not in the document]
Comment (robobario) [reply impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java]: I went with using mirror.gcr.io as default, plus an env var `APACHE_KAFKA_IMAGE_REPO` allowing users to override, this follows the existing mechanisms.

I did wonder a little whether we should repurpose `KAFKA_IMAGE_REPO`. Like add a `LEGACY_KAFKA_IMAGE_REPO` variable and make a little hierarchy of preference like:

kafka <4.1: userDefinedVersion > LEGACY_KAFKA_IMAGE_REPO > KAFKA_IMAGE_REPO > default
kafka >=4.1: userDefinedVersion > APACHE_KAFKA_IMAGE_REPO > KAFKA_IMAGE_REPO > default

so that a user who just cares about the latest kafka version could use KAFKA_IMAGE_REPO, but I figure it's a bit much.

On that, maybe we should have a table of env vars on the front README.

**Comment by robobario** [comment]
I updated #364 to note that we will use the JRE images, we can keep that issue open to revisit the upstream native images in future

Also created a ticket to make testing Kafka Queues more convenient when it goes GA #510

**Comment by robobario** [comment]
and a successful run using container mode:

https://github.com/robobario/kroxylicious/actions/runs/17571164043/job/49907510188?pr=36

a bit of tweaking needed to fix the image mirror substitutor to not replace the `OneShotContainer` image from `*.redhat.com` (should we really be using `latest`?) and also a test needed an await added, I suspect just a little slower for a transaction to commit with the image version.
